ecosystem/sf-indexer/aptos-sf-indexer/src/indexer/substream_processor.rs

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    counters::{
+        GOT_CONNECTION, PROCESSOR_ERRORS, PROCESSOR_INVOCATIONS, PROCESSOR_SUCCESSES,
+        UNABLE_TO_GET_CONNECTION,
+    },
+    database::{execute_with_better_error, PgDbPool, PgPoolConnection},
+    indexer::{errors::BlockProcessingError, processing_result::ProcessingResult},
+    models::indexer_states::IndexerState,
+    proto::BlockScopedData,
+    schema,
+};
+use async_trait::async_trait;
+use diesel::{
+    sql_query,
+    sql_types::{BigInt, Text},
+    RunQueryDsl,
+};
+use schema::indexer_states::{self, dsl};
+use std::fmt::Debug;
+
+/// The `SubstreamProcessor` processes the output from a substream specific to the server instance
+#[async_trait]
+pub trait SubstreamProcessor: Send + Sync + Debug {
+    /// name of the substream module, for status logging
+    /// This will get stored in the database for each (substream_module, block_height) pair
+    fn substream_module_name(&self) -> &'static str;
+
+    /// Accepts a transaction, and processes it. This method will be called from `process_substream_with_status`
+    /// In case a transaction cannot be processed, returns an error: the processor will mark it as failed in the database,
+    /// and it will be retried next time the indexer is started.
+    async fn process_substream(
+        &self,
+        stream_data: BlockScopedData,
+        block_height: u64,
+    ) -> Result<ProcessingResult, BlockProcessingError>;
+
+    /// Gets a reference to the connection pool
+    /// This is used by the `get_conn()` helper below
+    fn connection_pool(&self) -> &PgDbPool;
+
+    //* Below are helper methods that don't need to be implemented *//
+
+    /// Gets the connection.
+    /// If it was unable to do so (default timeout: 30s), it will keep retrying until it can.
+    /// It's a static method because we need the connection before the processor is initialized
+    fn get_conn(pool: &PgDbPool) -> PgPoolConnection {
```

Comment:
lets move static methods to the top to avoid interspersing

Reply:
This format is copied from the existing transaction_processors. I think that your initial logic is to put functions that need to be implemented above and functions that don't below. I kinda like that but up to you.

Reply:
oi lol u right